core/src/Streamly/Internal/Data/MutArray/Type.hs

```diff
@@ -2906,21 +2933,53 @@ cast arr =
 -- /Pre-release/
 --
 {-# INLINE unsafePinnedAsPtr #-}
-unsafePinnedAsPtr :: MonadIO m => MutArray a -> (Ptr a -> m b) -> m b
+unsafePinnedAsPtr :: MonadIO m => MutArray a -> (Ptr a -> Int -> m b) -> m b
 unsafePinnedAsPtr arr f =
     Unboxed.unsafePinnedAsPtr
-        (arrContents arr) (\ptr -> f (ptr `plusPtr` arrStart arr))
+        (arrContents arr)
+        (\ptr -> f (ptr `plusPtr` arrStart arr) (byteLength arr))
 
 {-# DEPRECATED asPtrUnsafe "Please use unsafePinnedAsPtr instead." #-}
 {-# INLINE asPtrUnsafe #-}
 asPtrUnsafe :: MonadIO m => MutArray a -> (Ptr a -> m b) -> m b
-asPtrUnsafe = unsafePinnedAsPtr
+asPtrUnsafe a f = unsafePinnedAsPtr a (\p _ -> f p)
 
+-- NOTE: unsafeAsPtr is safe to use unsafe with ffi given that the ffi function
+-- does not need the pointer to be valid after the call has completed.
 {-# INLINE unsafeAsPtr #-}
-unsafeAsPtr :: MonadIO m => MutArray a -> (Ptr a -> m b) -> m b
+unsafeAsPtr :: MonadIO m => MutArray a -> (Ptr a -> Int -> m b) -> m b
 unsafeAsPtr arr f =
     Unboxed.unsafeAsPtr
-        (arrContents arr) (\ptr -> f (ptr `plusPtr` arrStart arr))
+        (arrContents arr)
+        (\ptr -> f (ptr `plusPtr` arrStart arr) (byteLength arr))
+
+-- | @unsafeCreateUsingPtr capacity populater@ creates an array of @capacity@
+-- bytes lets the @populater@ function populate it. The @populater@ get the
+-- pointer to the array and should return the amount of the capacity populated
+-- in bytes.
+--
+-- /Unsafe/ because the pointer given should be used with care. Bytes populated
+-- should be lesser than the total capacity.
+{-# INLINE unsafeCreateUsingPtr #-}
+unsafeCreateUsingPtr
+    :: MonadIO m => Int -> (Ptr Word8 -> m Int) -> m (MutArray Word8)
+unsafeCreateUsingPtr cap pop = do
+    (arr :: MutArray Word8) <- emptyOf cap
+    len <- Unboxed.unsafeAsPtr (arrContents arr) pop
+    when (len > cap) (error "unsafeCreateUsingPtr: length > capacity")
```

Comment:
It's always good to print as much relevant info as we can e.g. the length and capacity here.